File: apps/api/app/integrations.py
from datetime import timedelta
from typing import Any

import stripe
from google.cloud import storage, tasks_v2

from app.firestore_store import utc_now
from app.settings import settings


ONE_OFF_PRODUCTS = {
    "text_once": {"name": "Text once", "amount_cents": 699, "duration_minutes": 0, "type": "text"},
    "quick_call": {"name": "Quick Call", "amount_cents": 299, "duration_minutes": 3, "type": "call"},
    "standard_call": {"name": "Standard Call", "amount_cents": 699, "duration_minutes": 25, "type": "call"},
    "long_call": {"name": "Long Call", "amount_cents": 1499, "duration_minutes": 60, "type": "call"},
}

SUBSCRIPTION_PRODUCTS = {
    "text_friend": {"name": "Text Friend", "amount_cents": 499},
    "friend": {"name": "Friend", "amount_cents": 1900},
    "close_friend": {"name": "Close Friend", "amount_cents": 2999},
    "always_there": {"name": "Always There", "amount_cents": 4900},
}


def provider_configured(value: str | None) -> bool:
    return bool(value and value.strip() and not value.startswith("replace-me"))


class StripeClient:
    def __init__(self) -> None:
        self.configured = provider_configured(settings.stripe_secret_key)
        if self.configured:
            stripe.api_key = settings.stripe_secret_key

    def create_one_off_checkout(
        self, *, session_id: str, product_key: str, priority_bid_cents: int
    ) -> dict[str, Any]:
        product = ONE_OFF_PRODUCTS[product_key]
        total_cents = product["amount_cents"] + priority_bid_cents
        if not self.configured:
            return {
                "configured": False,
                "checkout_url": f"{settings.app_base_url}/queue?session={session_id}&preview=1",
                "stripe_session_id": f"preview_{session_id}",
            }

        checkout = stripe.checkout.Session.create(
            mode="payment",
            success_url=f"{settings.app_base_url}/queue?session={session_id}",
            cancel_url=f"{settings.app_base_url}/start?cancelled=1",
            client_reference_id=session_id,
            metadata={"session_id": session_id, "product": product_key},
            line_items=[
                {
                    "quantity": 1,
                    "price_data": {
                        "currency": "usd",
                        "unit_amount": total_cents,
                        "product_data": {"name": product["name"]},
                    },
                }
            ],
        )
        return {
            "configured": True,
            "checkout_url": checkout.url,
            "stripe_session_id": checkout.id,
        }

    def create_subscription_checkout(self, *, user_id: str, plan_key: str) -> dict[str, Any]:
        product = SUBSCRIPTION_PRODUCTS[plan_key]
        if not self.configured:
            return {
                "configured": False,
                "checkout_url": f"{settings.app_base_url}/account?plan={plan_key}&preview=1",
                "stripe_session_id": f"preview_subscription_{user_id}_{plan_key}",
            }

        checkout = stripe.checkout.Session.create(
            mode="subscription",
            success_url=f"{settings.app_base_url}/account?subscription=success",
            cancel_url=f"{settings.app_base_url}/start?cancelled=1",
            client_reference_id=user_id,
            metadata={"user_id": user_id, "plan": plan_key},
            line_items=[
                {
                    "quantity": 1,
                    "price_data": {
                        "currency": "usd",
                        "unit_amount": product["amount_cents"],
                        "recurring": {"interval": "month"},
                        "product_data": {"name": product["name"]},
                    },
                }
            ],
        )
        return {
            "configured": True,
            "checkout_url": checkout.url,
            "stripe_session_id": checkout.id,
        }


class CallsClient:
    @property
    def configured(self) -> bool:
        return all(
            [
                provider_configured(settings.livekit_url),
                provider_configured(settings.livekit_api_key),
                provider_configured(settings.livekit_api_secret),
            ]
        )

    def create_room(self, *, session_id: str, user_id: str) -> dict[str, Any]:
        # Token signing is added when the LiveKit/Daily secret is installed.
        return {
            "configured": self.configured,
            "provider": "livekit",
            "room_name": f"pickup-{session_id}",
            "join_url": f"{settings.app_base_url}/call?session={session_id}",
            "token": None if not self.configured else f"token-required-for-{user_id}",
        }


class StorageClient:
    @property
    def configured(self) -> bool:
        return provider_configured(settings.google_cloud_storage_bucket)

    def signed_recording_url(self, *, object_name: str) -> str | None:
        if not self.configured:
            return None
        client = storage.Client(project=settings.google_cloud_project)
        blob = client.bucket(settings.google_cloud_storage_bucket).blob(object_name)
        return blob.generate_signed_url(
            version="v4",
            expiration=timedelta(minutes=15),
            method="PUT",
            content_type="application/octet-stream",
        )


class TasksClient:
    @property
    def configured(self) -> bool:
        return bool(settings.google_cloud_project and settings.cloud_tasks_queue)

    def enqueue_session_timeout(self, *, session_id: str, delay_seconds: int) -> dict[str, Any]:
        if not self.configured:
            return {"configured": False, "task_name": None}
        client = tasks_v2.CloudTasksClient()
        parent = client.queue_path(
            settings.google_cloud_project,
            settings.cloud_tasks_location,
            settings.cloud_tasks_queue,
        )
        task = {
            "schedule_time": utc_now() + timedelta(seconds=delay_seconds),
            "http_request": {
                "http_method": tasks_v2.HttpMethod.POST,
                "url": f"{settings.app_base_url}/api/internal/session-timeout/{session_id}",
            },
        }
        created = client.create_task(parent=parent, task=task)
        return {"configured": True, "task_name": created.name}


stripe_client = StripeClient()
calls_client = CallsClient()
storage_client = StorageClient()
tasks_client = TasksClient()

File: apps/web/server/integrations.ts
import { Storage } from "@google-cloud/storage";
import { CloudTasksClient } from "@google-cloud/tasks";
import { AccessToken, TrackSource, type VideoGrant } from "livekit-server-sdk";
import Stripe from "stripe";
import { settings } from "./settings";
import { utcNow } from "./store";

export const ONE_OFF_PRODUCTS = {
  text_once: { name: "Text once", amount_cents: 699, duration_minutes: 0, type: "text" },
  quick_call: { name: "Quick Call", amount_cents: 299, duration_minutes: 3, type: "call" },
  standard_call: { name: "Standard Call", amount_cents: 699, duration_minutes: 25, type: "call" },
  long_call: { name: "Long Call", amount_cents: 1499, duration_minutes: 60, type: "call" }
} as const;

export const SUBSCRIPTION_PRODUCTS = {
  text_friend: { name: "Text Friend", amount_cents: 499 },
  friend: { name: "Friend", amount_cents: 1900 },
  close_friend: { name: "Close Friend", amount_cents: 2999 },
  always_there: { name: "Always There", amount_cents: 4900 }
} as const;

export type OneOffProductKey = keyof typeof ONE_OFF_PRODUCTS;
export type SubscriptionPlanKey = keyof typeof SUBSCRIPTION_PRODUCTS;

function providerConfigured(value: string | undefined): boolean {
  return Boolean(value && value.trim() && !value.startsWith("replace-me"));
}

export type CheckoutResult = {
  configured: boolean;
  checkout_url: string;
  stripe_session_id: string;
};

export const stripeClient = {
  get configured(): boolean {
    return providerConfigured(settings.stripeSecretKey);
  },

  client(): Stripe {
    return new Stripe(settings.stripeSecretKey!);
  },

  async createOneOffCheckout(options: {
    sessionId: string;
    productKey: OneOffProductKey;
    priorityBidCents: number;
  }): Promise<CheckoutResult> {
    const { sessionId, productKey, priorityBidCents } = options;
    const product = ONE_OFF_PRODUCTS[productKey];
    const totalCents = product.amount_cents + priorityBidCents;
    if (!this.configured) {
      return {
        configured: false,
        checkout_url: `${settings.appBaseUrl}/book?session=${sessionId}&preview=1`,
        stripe_session_id: `preview_${sessionId}`
      };
    }

    const checkout = await this.client().checkout.sessions.create({
      mode: "payment",
      success_url: `${settings.appBaseUrl}/book?session=${sessionId}`,
      cancel_url: `${settings.appBaseUrl}/start?cancelled=1`,
      client_reference_id: sessionId,
      metadata: { session_id: sessionId, product: productKey },
      line_items: [
        {
          quantity: 1,
          price_data: {
            currency: "usd",
            unit_amount: totalCents,
            product_data: { name: product.name }
          }
        }
      ]
    });
    return {
      configured: true,
      checkout_url: checkout.url ?? "",
      stripe_session_id: checkout.id
    };
  },

  async createSubscriptionCheckout(options: {
    userId: string;
    planKey: SubscriptionPlanKey;
  }): Promise<CheckoutResult> {
    const { userId, planKey } = options;
    const product = SUBSCRIPTION_PRODUCTS[planKey];
    if (!this.configured) {
      return {
        configured: false,
        checkout_url: `${settings.appBaseUrl}/start?plan=${planKey}&activated=1`,
        stripe_session_id: `preview_subscription_${userId}_${planKey}`
      };
    }

    const checkout = await this.client().checkout.sessions.create({
      mode: "subscription",
      success_url: `${settings.appBaseUrl}/account?subscription=success`,
      cancel_url: `${settings.appBaseUrl}/start?cancelled=1`,
      client_reference_id: userId,
      metadata: { user_id: userId, plan: planKey },
      line_items: [
        {
          quantity: 1,
          price_data: {
            currency: "usd",
            unit_amount: product.amount_cents,
            recurring: { interval: "month" },
            product_data: { name: product.name }
          }
        }
      ]
    });
    return {
      configured: true,
      checkout_url: checkout.url ?? "",
      stripe_session_id: checkout.id
    };
  }
};

export type CallRole = "caller" | "host";

export type RoomResult = {
  configured: boolean;
  provider: "livekit";
  livekit_url: string | null;
  room_name: string;
  join_url: string;
  token: string | null;
};

export const callsClient = {
  get configured(): boolean {
    return (
      providerConfigured(settings.livekitUrl) &&
      providerConfigured(settings.livekitApiKey) &&
      providerConfigured(settings.livekitApiSecret)
    );
  },

  roomName(sessionId: string): string {
    return `callsomeone-${sessionId}`;
  },

  async accessToken(options: {
    sessionId: string;
    identity: string;
    displayName: string;
    role: CallRole;
  }): Promise<string> {
    const { sessionId, identity, displayName, role } = options;
    const grants: VideoGrant = {
      roomJoin: true,
      room: this.roomName(sessionId),
      canPublish: true,
      canSubscribe: true,
      canPublishSources: [TrackSource.MICROPHONE],
      roomAdmin: role === "host"
    };
    const token = new AccessToken(settings.livekitApiKey, settings.livekitApiSecret, {
      identity,
      name: displayName
    });
    token.addGrant(grants);
    return token.toJwt();
  },

  async createRoom(options: {
    sessionId: string;
    userId: string;
    role?: CallRole;
  }): Promise<RoomResult> {
    const { sessionId, userId, role = "caller" } = options;
    const roomName = this.roomName(sessionId);
    let token: string | null = null;
    if (this.configured) {
      const identity = role === "host" ? settings.hostUserId : userId;
      const displayName = role === "host" ? "Host" : "Guest";
      token = await this.accessToken({ sessionId, identity, displayName, role });
    }

    return {
      configured: this.configured,
      provider: "livekit",
      livekit_url: settings.livekitUrl ?? null,
      room_name: roomName,
      join_url: `${settings.appBaseUrl}/call?session=${sessionId}`,
      token
    };
  }
};

export const storageClient = {
  get configured(): boolean {
    return providerConfigured(settings.googleCloudStorageBucket);
  },

  async signedRecordingUrl(objectName: string): Promise<string | null> {
    if (!this.configured) return null;
    const storage = new Storage({ projectId: settings.googleCloudProject });
    const [url] = await storage
      .bucket(settings.googleCloudStorageBucket!)
      .file(objectName)
      .getSignedUrl({
        version: "v4",
        expires: Date.now() + 15 * 60 * 1000,
        action: "write",
        contentType: "application/octet-stream"
      });
    return url;
  }
};

export const tasksClient = {
  get configured(): boolean {
    return Boolean(settings.googleCloudProject && settings.cloudTasksQueue);
  },

  async enqueueSessionTimeout(options: {
    sessionId: string;
    delaySeconds: number;
  }): Promise<{ configured: boolean; task_name: string | null }> {
    const { sessionId, delaySeconds } = options;
    if (!this.configured) return { configured: false, task_name: null };
    const client = new CloudTasksClient();
    const parent = client.queuePath(
      settings.googleCloudProject!,
      settings.cloudTasksLocation,
      settings.cloudTasksQueue
    );
    const scheduleTime = new Date(utcNow().getTime() + delaySeconds * 1000);
    const [created] = await client.createTask({
      parent,
      task: {
        scheduleTime: { seconds: Math.floor(scheduleTime.getTime() / 1000) },
        httpRequest: {
          httpMethod: "POST",
          url: `${settings.appBaseUrl}/api/internal/session-timeout/${sessionId}`
        }
      }
    });
    return { configured: true, task_name: created.name ?? null };
  }
};
